File: web/flood-fill.js
// InkFrame — scanline flood fill
// -----------------------------------------------------------------------------
// Pure algorithm, no DOM. A 1:1 port of core-common/src/main/kotlin/com/inkframe/
// core/common/FloodFill.kt so bug fixes port back trivially in either direction
// (same pattern established by gif-encoder.js).
//
// Uses the classic span-based (scanline) algorithm: for each seed, fill the
// contiguous horizontal run, then queue the rows above and below. This is far
// cheaper than a naive 4-way recursion and won't blow the JS call stack on
// large regions (2048x2048 fills stay comfortably under a few ms typical).
//
// Matching uses a per-channel tolerance against the ORIGINAL colour at the
// seed, so anti-aliased edges can be included by raising `tolerance`. Already-
// target pixels are treated as a no-op to avoid infinite loops when
// fill === target.
//
// The important pixel-format detail: on little-endian machines (which is every
// browser + Android WebView), a canvas ImageData's Uint32Array view stores
// pixels as 0xAABBGGRR (native byte order). This module doesn't care about the
// channel order -- it just compares integers -- so both callers and this file
// stay format-agnostic. Pack your fill colour the same way you read pixels.
'use strict';

function floodFill(argb, width, height, seedX, seedY, fillArgb, tolerance) {
  if (argb.length < width * height) throw new Error(`argb too small for ${width}x${height}`);
  if (seedX < 0 || seedY < 0 || seedX >= width || seedY >= height) {
    return { changed:false,minX:0,minY:0,maxX:0,maxY:0,pixelsFilled:0 };
  }
  fillArgb = fillArgb >>> 0;
  const target = argb[seedY * width + seedX];
  if (target === fillArgb) return { changed:false,minX:0,minY:0,maxX:0,maxY:0,pixelsFilled:0 };
  const tol = Math.max(0, Math.min(255, tolerance | 0));
  let matches;
  if (tol === 0) {
    matches = c => c !== fillArgb && c === target;
  } else {
    const ta=(target>>>24)&0xFF,tr=(target>>>16)&0xFF,tg=(target>>>8)&0xFF,tb=target&0xFF;
    matches = c => {
      if (c === fillArgb) return false;
      const da=((c>>>24)&0xFF)-ta,dr=((c>>>16)&0xFF)-tr,dg=((c>>>8)&0xFF)-tg,db=(c&0xFF)-tb;
      return Math.abs(da)<=tol&&Math.abs(dr)<=tol&&Math.abs(dg)<=tol&&Math.abs(db)<=tol;
    };
  }
  let minX=seedX,minY=seedY,maxX=seedX,maxY=seedY,filled=0;
  const sx=[seedX],sy=[seedY];
  while (sx.length) {
    const px=sx.pop(),py=sy.pop(),rowBase=py*width;
    if (!matches(argb[rowBase+px])) continue;
    let left=px; while(left-1>=0&&matches(argb[rowBase+left-1])) left--;
    let right=px; while(right+1<width&&matches(argb[rowBase+right+1])) right++;
    for(let x=left;x<=right;x++){argb[rowBase+x]=fillArgb;filled++;}
    if(left<minX)minX=left;if(right>maxX)maxX=right;if(py<minY)minY=py;if(py>maxY)maxY=py;
    queueRow(argb,width,left,right,py-1,matches,sx,sy);
    queueRow(argb,width,left,right,py+1,matches,sx,sy);
  }
  return { changed:filled>0,minX,minY,maxX,maxY,pixelsFilled:filled };
}

function queueRow(argb,width,left,right,y,matches,sx,sy){
  if(y<0)return;
  const rowBase=y*width;if(rowBase>=argb.length)return;
  let x=left;
  while(x<=right){
    while(x<=right&&!matches(argb[rowBase+x]))x++;
    if(x>right)break;
    sx.push(x);sy.push(y);
    while(x<=right&&matches(argb[rowBase+x]))x++;
  }
}

{
  const _api={floodFill};
  if(typeof window!=='undefined')window.InkFrameFloodFill=_api;
  if(typeof module!=='undefined'&&module.exports)module.exports=_api;
}

// This sibling module is the last external script before the app IIFE. Load the
// velocity runtime synchronously here so it can wrap the canvas pointer listener
// before that listener is registered. jsdom is skipped because the smoke test
// has no HTTP server; brush-dynamics has its own Node regression suite.
if(typeof document!=='undefined'&&typeof window!=='undefined'&&document.readyState==='loading'&&
   !window.InkFrameBrushDynamics&&
   !/jsdom/i.test((window.navigator&&window.navigator.userAgent)||'')){
  document.write('<script src="brush-dynamics.js"><\/script>');
}

File: FloodFill.kt
package com.inkframe.core.common

/**
 * Pure scanline flood fill over a top-down ARGB int array. No GL / Android — the
 * engine supplies pixels read back from a cel surface, runs this, and writes the result
 * back. Fully unit-tested.
 *
 * Uses the classic span-based (scanline) algorithm: for each seed, fill the contiguous
 * horizontal run, then queue the rows above and below. This is far cheaper than a
 * naive 4-way recursion and won't overflow the stack on large regions.
 *
 * Matching uses a per-channel tolerance against the *original* colour at the seed, so
 * anti-aliased edges can be included by raising [tolerance]. Already-target pixels are
 * treated as a no-op to avoid infinite loops when fill == target.
 */
object FloodFill {

    /** Result of a fill: which pixels changed (as a dirty rect) and how many. */
    data class FillResult(
        val changed: Boolean,
        val minX: Int,
        val minY: Int,
        val maxX: Int,
        val maxY: Int,
        val pixelsFilled: Int,
    ) {
        /** Inclusive dirty rectangle as an [IntRect], or null if nothing changed. */
        fun dirtyRect(): IntRect? =
            if (!changed) null else IntRect(minX, minY, maxX - minX + 1, maxY - minY + 1)
    }

    /**
     * Flood-fills [argb] (modified in place) starting at ([seedX], [seedY]) with
     * [fillArgb], replacing the connected region whose colour matches the seed's within
     * [tolerance] (0..255 per channel, including alpha).
     *
     * @return a [FillResult] describing the affected area.
     */
    fun fill(
        argb: IntArray,
        width: Int,
        height: Int,
        seedX: Int,
        seedY: Int,
        fillArgb: Int,
        tolerance: Int = 0,
    ): FillResult {
        require(argb.size >= width * height) { "argb too small for ${width}x$height" }
        if (seedX < 0 || seedY < 0 || seedX >= width || seedY >= height) {
            return FillResult(false, 0, 0, 0, 0, 0)
        }

        val target = argb[seedY * width + seedX]
        // No-op if the seed already equals the fill colour (avoids spinning forever).
        if (target == fillArgb) return FillResult(false, 0, 0, 0, 0, 0)

        val tol = tolerance.coerceIn(0, 255)
        fun matches(c: Int): Boolean {
            if (c == fillArgb) return false // already filled this pass
            if (tol == 0) return c == target
            val da = kotlin.math.abs(((c ushr 24) and 0xFF) - ((target ushr 24) and 0xFF))
            val dr = kotlin.math.abs(((c ushr 16) and 0xFF) - ((target ushr 16) and 0xFF))
            val dg = kotlin.math.abs(((c ushr 8) and 0xFF) - ((target ushr 8) and 0xFF))
            val db = kotlin.math.abs((c and 0xFF) - (target and 0xFF))
            return da <= tol && dr <= tol && dg <= tol && db <= tol
        }

        var minX = seedX; var minY = seedY; var maxX = seedX; var maxY = seedY
        var filled = 0

        // Stack of seed points; each expands into a horizontal span.
        val stackX = ArrayDeque<Int>()
        val stackY = ArrayDeque<Int>()
        stackX.addLast(seedX); stackY.addLast(seedY)

        while (stackX.isNotEmpty()) {
            val px = stackX.removeLast()
            val py = stackY.removeLast()
            val rowBase = py * width

            // Skip if this pixel was already handled.
            if (!matches(argb[rowBase + px])) continue

            // Expand left and right to find the span bounds.
            var left = px
            while (left - 1 >= 0 && matches(argb[rowBase + left - 1])) left--
            var right = px
            while (right + 1 < width && matches(argb[rowBase + right + 1])) right++

            // Fill the span and record dirty bounds.
            for (x in left..right) {
                argb[rowBase + x] = fillArgb
                filled++
            }
            if (left < minX) minX = left
            if (right > maxX) maxX = right
            if (py < minY) minY = py
            if (py > maxY) maxY = py

            // Queue spans above and below: add a seed wherever the neighbour row matches.
            queueRow(argb, width, left, right, py - 1, ::matches, stackX, stackY)
            queueRow(argb, width, left, right, py + 1, ::matches, stackX, stackY)
        }

        return FillResult(filled > 0, minX, minY, maxX, maxY, filled)
    }

    /** Adds one seed per contiguous matching run in row [y] within [left]..[right]. */
    private inline fun queueRow(
        argb: IntArray,
        width: Int,
        left: Int,
        right: Int,
        y: Int,
        matches: (Int) -> Boolean,
        stackX: ArrayDeque<Int>,
        stackY: ArrayDeque<Int>,
    ) {
        if (y < 0) return
        val rowBase = y * width
        if (rowBase >= argb.size) return
        var x = left
        while (x <= right) {
            // advance to the next matching pixel
            while (x <= right && !matches(argb[rowBase + x])) x++
            if (x > right) break
            // one seed for this run, then skip to its end
            stackX.addLast(x); stackY.addLast(y)
            while (x <= right && matches(argb[rowBase + x])) x++
        }
    }
}
